src/charm.py

```diff
@@ -558,6 +535,116 @@ def configuration_check(self) -> bool:
             logger.exception("Invalid configuration")
             return False
 
+    def get_node_hosts(self) -> set[str]:
+        """Return the node ports of nodes where units of this app are scheduled."""
+        peer_relation = self.model.get_relation(PEER_RELATION_NAME)
+        if not peer_relation:
+            return set()
+
+        def _get_node_address(node) -> str:
+            # OpenStack will return an internal hostname, not externally accessible
+            # Preference: ExternalIP > InternalIP > Hostname
+            for typ in ["ExternalIP", "InternalIP", "Hostname"]:
+                for address in node.status.addresses:
+                    if address.type == typ:
+                        return address.address
```

Comment:
I think we can move this outside of `get_node_hosts()`

Reply:
addressed in https://github.com/canonical/pgbouncer-k8s-operator/commit/46b037d599cbddcda6fdaed26488c21bff6dc6d6